daemon/container_operations_unix.go

```diff
@@ -625,11 +690,15 @@ func (daemon *Daemon) ConnectToNetwork(container *container.Container, idOrName
 	return nil
 }
 
-func (daemon *Daemon) connectToNetwork(container *container.Container, idOrName string, updateSettings bool) (err error) {
+func (daemon *Daemon) connectToNetwork(container *container.Container, idOrName string, endpointConfig *networktypes.EndpointSettings, updateSettings bool) (err error) {
 	if container.HostConfig.NetworkMode.IsContainer() {
 		return runconfig.ErrConflictSharedNetwork
 	}
 
+	if !container.HostConfig.NetworkMode.IsUserDefined() && networktypes.HasUserDefinedIPAddress(endpointConfig) {
```

Comment:
adding to @tonistiigi's comment, I think it is incorrect to check for HostConfig.NetworkMode here.
it must be the network that is being connected to using `containertypes.NetworkMode(idOrName).IsUserDefined`.

If we are going down this path, then I think it is proper to also check if the network has a configured `subnet`.
The user should not be depending on the auto-generated subnet and choose preferred-ip. It can cause the same container restart issues.